File: ctci_1_1.swift
#!/usr/bin/env swift
// cracking the coding interview q 1.1
// determine if a string has all unique characters

func unique(n: String) -> Bool {
  let charactersAsSet = Array(Set(Array(n))).sorted { $0 < $1 }
  let charactersAsArray = Array(n).sorted { $0 < $1 }
  return charactersAsSet == charactersAsArray
}

println(unique("abcdefgh "))
println(unique("hello world"))

func uniqueTwo(n: String) -> Bool {
  var chars: [Character] = []
  for index in 0..<count(n) {
    let c = n[advance(n.startIndex,index)]
    if contains(chars,c) {
      return false
    } else {
      chars.append(c)
    }
  }
  return true
}

println(uniqueTwo("abcdefgh "))
println(uniqueTwo("hello world"))

File: ctci_1_1.js
#!/usr/bin/env node

function arrayContains(list,e) {
  return list.filter(function(a) { return a === e }).length != 0
}

function unique(s) {
  var charsArray = s.split("");
  var charsSet = [];
  for (var i in charsArray) {
    if (!arrayContains(charsSet,charsArray[i])) {
      charsSet.push(charsArray[i]);
    }
  }
  return charsArray.length === charsSet.length
}

console.log(unique("abc"));
console.log(unique("abba"));
console.log(unique("hello world"));
console.log(unique("abcdefghijkl"));
